pkg/providerconfig/types.go

```diff
@@ -67,6 +67,56 @@ type ConfigVarString struct {
 // causing a recursion
 type configVarStringWithoutUnmarshaller ConfigVarString
 
+// This is done to not have the json object cluttered with empty strings
+// This will eventually hopefully be resolved within golang itself
+// https://github.com/golang/go/issues/11939
+func (configVarString ConfigVarString) MarshalJSON() ([]byte, error) {
+	var secretKeyRefEmpty, configMapKeyRefEmpty bool
+	if configVarString.SecretKeyRef.ObjectReference.Namespace == "" &&
+		configVarString.SecretKeyRef.ObjectReference.Name == "" &&
+		configVarString.SecretKeyRef.Key == "" {
+		secretKeyRefEmpty = true
+	}
+
+	if configVarString.ConfigMapKeyRef.ObjectReference.Namespace == "" &&
+		configVarString.ConfigMapKeyRef.ObjectReference.Name == "" &&
+		configVarString.ConfigMapKeyRef.Key == "" {
+		configMapKeyRefEmpty = true
+	}
+
+	if secretKeyRefEmpty && configMapKeyRefEmpty {
```

Comment:
Maybe rename the variables to something like `isSecretKeyRef` ?
I find it a bit more intuitive having true-focused variables instead of variables which are `true` when a condition is false.
Also it might make the following code a bit more readable.

```go
if !secretKeyRefEmpty {
    // handle secret ref 
}
```
vs.
```go
if isSecretKeyRef {
    // handle secret ref 
}
```


Though is it 100% personal preference. Not a request for change

Reply:
Applies to `ConfigVarBool` as well

Reply:
Good suggestions thought, I'll change that later